rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,345 @@
+---
+feature: sign-commits
+start-date: 2021-08-14
+author: Las Safin
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues: https://github.com/NixOS/rfcs/pull/34
+---
+
+# Summary
+[summary]: #summary
+
+- Add a way for flake inputs that point to Git repositories to be verified for trust.
+- Speed up fetching Git repositories with Nix.
+- Recommend and eventually require that people who commit directly to
+  https://github.com/NixOS/nixpkgs and https://github.com/NixOS/nix,
+  so called authorized committers, sign their commits.
+  - Commits inside merges do not have to be signed as long as the merge commit itself is signed,
+		such that only Committers have to sign their commits.
+
+# Motivation
+[motivation]: #motivation
+
+Signing commits would allow cryptographically verifying that the source code
+hasn't been tampered by an untrusted party.
+
+This would ensure that when pulling from GitHub, or some other source,
+you can check if the source code has been signed by a trusted person.
+
+This is already done by Guix:
+- https://guix.gnu.org/en/blog/2020/securing-updates/
+- https://github.com/guix-mirror/guix/commits/master
+- https://guix.gnu.org/manual/en/html_node/Commit-Access.html
+- https://guix.gnu.org/manual/en/html_node/Channel-Authentication.html
+
+It is also done by most conventional distros, however, they sign
+the output from the compilation rather than the source code, since
+they distribute the output from the compilation.
+
+This change would also allow using 3rd-party mirrors of Nixpkgs,
+since we can verify the source code independent of where it came from.
+
+This RFC was motivated by work during Summer of Nix, specifically,
+the Libre SoC maintainer requested that source integrity be
+maintained.
+
+# Detailed design
+[design]: #detailed-design
+
+## Signing the commits
+[signing]: #signing-the-commits
+
+The following will be required from 2022-03-01T00:00Z on,
+but is strongly recommended from now on.
+
+Currently merging is done through GitHub, which would not work in this case.
+Merging would have to happen outside of GitHub, so that you can sign the commits,
+since currently the API doesn't support signing commits with your own PGP
+key: https://github.com/cli/cli/issues/1318.
+
+We would make a script for merging pull requests instead, something like:
+```bash
+#!/bin/sh
+# nixpkgsmerge
+
+git switch master
+git pull
+gh pr checkout $1
+git switch master
+git merge -m "Merged #$1" --no-ff --gpg-sign -
+git push origin master
+gh pr close $1
+```
+Then
+```bash
+nixpkgsmerge 123
+```
+
+For mobile platforms, you would use [nix-on-droid](https://github.com/t184256/nix-on-droid)
+to run the script instead.
+
+A merge commit must always be made so that there is something to sign.
+
+A git hook like the one presented here is also recommended,
+so that unsigned commits aren't pushed by accident.
+This would be put in `/.git/hooks/pre-push` of the repository to
```

Comment:
Yes, this is very true.

Reply:
I don't want any hook for now that verifies this until someone has proven that this works. Git hooks are incredible anyoing if they are not functioning properly.

Reply:
I'm already using this hook, but for other repositories. It's quite simple and doesn't really fail, since it only checks the latest commit. If you wanted to disable it you'd just push with `--no-verify`.